File: Sources/Forage/Recipe/Pagination.swift
import Foundation

/// HTTP-engine pagination strategies (the catch-all browser-engine
/// pagination is `BrowserPaginate`, which lives in its own module file).
///
/// Each variant declares the request-side param + response-side paths the
/// engine needs to drive the loop and recognize termination.
public enum Pagination: Hashable, Sendable {
    /// Send a page-number param (1-based or 0-based per `pageZeroIndexed`),
    /// response carries an `items` list and a `total` count. Stop when
    /// accumulated items ≥ total. Sweed.
    case pageWithTotal(
        itemsPath: PathExpr,
        totalPath: PathExpr,
        pageParam: String,
        pageSize: Int,
        pageZeroIndexed: Bool = false
    )

    /// Send a page-number param, response carries an `items` list. Stop when
    /// items shorter than page-size or empty. Leafbridge.
    case untilEmpty(
        itemsPath: PathExpr,
        pageParam: String,
        pageZeroIndexed: Bool = false
    )

    /// Server returns a continuation token in each response that gets sent
    /// back on the next request. Empty/nil cursor terminates.
    case cursor(
        itemsPath: PathExpr,
        cursorPath: PathExpr,
        cursorParam: String
    )
}

File: crates/forage-core/src/ast/pagination.rs
//! HTTP-engine pagination strategies.
//!
//! Browser-engine pagination is in `ast::browser`.

use serde::{Deserialize, Serialize};

use crate::ast::expr::PathExpr;

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub enum Pagination {
    /// Send a page-number param; response carries `items` + `total`. Stop
    /// when accumulated items ≥ total. (Sweed.)
    PageWithTotal {
        items_path: PathExpr,
        total_path: PathExpr,
        page_param: String,
        page_size: u32,
        #[serde(default)]
        page_zero_indexed: bool,
    },
    /// Send a page-number param; response carries `items`. Stop when items
    /// shorter than page-size or empty. (Leafbridge.)
    UntilEmpty {
        items_path: PathExpr,
        page_param: String,
        #[serde(default)]
        page_zero_indexed: bool,
    },
    /// Server returns a continuation token; empty/nil cursor terminates.
    Cursor {
        items_path: PathExpr,
        cursor_path: PathExpr,
        cursor_param: String,
    },
}
